README.md
## DPWalker

A Dependency-Walker like application for Linux

Based on: https://github.com/shentok/depends-ui

### What is it?

This is  a free utility that scans Linux ELF-formated binaries(Executable, Shared library) and builds a hierarchical tree diagram of all dependent libraries. 

### Detailed Information

For each dependency libraries found, it lists detailed information  about each libraries including a full path to the file. It recursively search information for each libraries.

"Export Symbol" Menu displays all the functions that are exported by current selected module. It reviles detailed information about searched symbols: base address, visibility, size, library name, etc... 

"Import Symbol" Menu displays and which of those are actually being imported by other libraries. 

### How to build?

You need binutils binary, Qt5 development package (which include libQt5Widgets, libQt5Gui, libQt5Core) for depencency. And it uses qmake for build.

For Ubuntu-based distros, you can install these dependency by following command:

```sh
sudo apt install build-essentials binutils qtbase5-dev qmake
```

Then, run qmake by following command:

```sh
mkdir build && cd build
qmake ../DPWalker.pro
```

And run GNU make for obtain binary

```shell
make
```

Or you can use Qt Creator to build this software.

### How to use?

When running software with no args, it pop-ups file-open dialog. 

Select target ELF-format file(*.so, executable file), it begins parsing information, and displays module information at left TreeView widget. 

You can view "Import Symbols", and "Export Symbols" information when you just click the Library name.

### Tested environment

This software tested on Ubuntu 18.04

### Screenshot

![](./MDResource/Screenshot.png)

ChangeLog
v 1.0.0 

- Initial version of this SW.

v 1.1.0

- Add spinner UI when listing is engaged.
- Change cursor shape when listing is engaged.





dpwalker/main.cpp
/***************************************************************************
 *   Copyright (C) 2007 by Bernhard Beschow                                *
 *   bbeschow (.at) cs.tu-berlin.de                                        *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <QApplication>
#include <QTimer>

#include "dependsui.h"


int main( int argc, char *argv[] )
{
	QApplication app( argc, argv );

	DependsUI ui;

	ui.show();

	if ( argc == 2 )
	{
		ui.openFile( argv[1] );
	}
	else
	{
        QTimer::singleShot( 0, &ui, SLOT(fileOpen()) );
	}

	return app.exec();
}


dpwalker/dependsui.h
/***************************************************************************
 *   Copyright (C) 2007 by Bernhard Beschow                                *
 *   bbeschow (.at) cs.tu-berlin.de                                        *
 *   Copyright (C) 2020 by C6H6Vita                                        *
 *   c6h6 (.at) outlook.kr                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef DEPENDSUI_H
#define DEPENDSUI_H

#include "ui_dependsuibase.h"

#include <QMainWindow>

class QTreeWidgetItem;

class DependsUI : public QMainWindow
{
Q_OBJECT

public:
	DependsUI( QWidget *parent = 0, Qt::WindowFlags flags = 0 );
	~DependsUI();
	/*$PUBLIC_FUNCTIONS$*/

	void openFile( const QString &file );

public slots:
	/*$PUBLIC_SLOTS$*/
	virtual void fileExit();
	virtual void fileOpen();

private slots:
	void dependencyExpanded( QTreeWidgetItem *pItem );
	void dependencySelected( QTreeWidgetItem *pItem );
	void setRootItemExpanded();
    void runCompleted();
    void aboutClicked();
private:
	Ui::DependsUIBase m_ui;
	QString m_dirPath;
    QMovie* gif;

    void setBusy(bool isBusy);
};

#endif



dpwalker/dependencyjob.h
/***************************************************************************
 *   Copyright (C) 2007 by Bernhard Beschow                                *
 *   bbeschow (.at) cs.tu-berlin.de                                        *
 *   Copyright (C) 2020 by C6H6Vita                                        *
 *   c6h6 (.at) outlook.kr                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#ifndef DEPENDENCYJOB_H
#define DEPENDENCYJOB_H

#include <qobject.h>

#include <QMap>
#include <QProcess>
#include <QTextStream>

class QTreeWidgetItem;

class LibraryPathsJob : public QObject
{
	Q_OBJECT
public:
	LibraryPathsJob( const QString &file, QMap<QString, QString> *libraryPaths, QObject *parent = 0 );

signals:
	void finished();

public slots:
	void start();

protected slots:
	void readLineStdout();

protected:
	QString m_file;
	QMap<QString, QString> *const m_libraryPaths;
	QProcess m_proc;
	QTextStream m_stream;
};

class DependenciesJob : public QObject
{
	Q_OBJECT
public:
	DependenciesJob( QTreeWidgetItem *treeWidgetItem, const QMap<QString, QString> *libraryPaths, QObject *parent = 0 );

signals:
	void finished();

public slots:
	void start();

protected slots:
	void readLineStdout();

protected:
	QTreeWidgetItem *const m_treeWidgetItem;
	const QMap<QString, QString> *m_libraryPaths;
	QProcess m_proc;
	QTextStream m_stream;
};

/**
 * @author Bernhard Beschow <bbeschow (.at) cs.tu-berlin.de>
 */
class DependencyJobs : public QObject
{
	Q_OBJECT
public:
	DependencyJobs( QTreeWidgetItem *treeWidgetItem, QObject *parent = 0 );

signals:
	void finished();

protected:
	QMap<QString, QString> m_libraryPaths;
	LibraryPathsJob m_libraryPathsJob;
	DependenciesJob m_dependenciesJob;
};

#endif


dpwalker/dependencyjob.cpp
/***************************************************************************
 *   Copyright (C) 2007 by Bernhard Beschow                                *
 *   bbeschow (.at) cs.tu-berlin.de                                        *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include "dependencyjob.h"

#include <QFileInfo>
#include <QRegExp>
#include <QTextStream>
#include <QTreeWidgetItem>

LibraryPathsJob::LibraryPathsJob( const QString &file, QMap<QString, QString> *libraryPaths, QObject *parent ) :
	QObject( parent ),
	m_file( file ),
	m_libraryPaths( libraryPaths ),
	m_proc(),
	m_stream( &m_proc )
{
	m_proc.setReadChannel( QProcess::StandardOutput );

	connect( &m_proc, SIGNAL(readyReadStandardOutput()), this, SLOT(readLineStdout()) );
	connect( &m_proc, SIGNAL(finished(int)), this, SLOT(readLineStdout()) );
}


void LibraryPathsJob::start()
{
	const QStringList args = QStringList() << m_file;

	m_proc.start( "ldd", args );
}


void LibraryPathsJob::readLineStdout()
{
	while ( m_proc.canReadLine() || !m_stream.atEnd() )
	{
		const QString line = m_stream.readLine();

		const QRegExp re( "^\\s+(\\S+) => (\\S+) \\(0x[0-9a-f]+\\)" );
		if ( re.indexIn( line ) >= 0 )
		{
			(*m_libraryPaths)[ re.cap( 1 ) ] = re.cap( 2 );
		}
	}

	if ( m_proc.state() == QProcess::NotRunning )
	{
		emit finished();
	}
}


DependenciesJob::DependenciesJob( QTreeWidgetItem *treeWidgetItem, const QMap<QString, QString> *libraryPaths, QObject *parent ) :
	QObject( parent ),
	m_treeWidgetItem( treeWidgetItem ),
	m_libraryPaths( libraryPaths ),
	m_proc(),
	m_stream( &m_proc )
{
	m_proc.setReadChannel( QProcess::StandardOutput );

	connect( &m_proc, SIGNAL(readyReadStandardOutput()), this, SLOT(readLineStdout()) );
	connect( &m_proc, SIGNAL(finished(int)), this, SLOT(readLineStdout()) );
}


void DependenciesJob::start()
{
	const QStringList args = QStringList() << "-p" << m_treeWidgetItem->text( 1 );

	m_proc.start( "objdump", args );
}


void DependenciesJob::readLineStdout()
{
	while ( m_proc.canReadLine() || !m_stream.atEnd() )
	{
		const QString line = m_stream.readLine();

		const QRegExp re( "\\s+NEEDED\\s+(\\S+)" );
		if ( re.indexIn( line ) >= 0 )
		{
			QTreeWidgetItem *pItem = new QTreeWidgetItem( m_treeWidgetItem );

			pItem->setText( 0, re.cap( 1 ) );
			pItem->setText( 1, (*m_libraryPaths)[ re.cap( 1 ) ] );
		}
	}

	if ( m_proc.state() == QProcess::NotRunning )
	{
		emit finished();
	}
}


DependencyJobs::DependencyJobs( QTreeWidgetItem *treeWidgetItem, QObject *parent ) :
	QObject( parent ),
	m_libraryPaths(),
	m_libraryPathsJob( treeWidgetItem->text( 1 ), &m_libraryPaths, this ),
	m_dependenciesJob( treeWidgetItem, &m_libraryPaths, this )
{
	connect( &m_libraryPathsJob, SIGNAL(finished()), &m_dependenciesJob, SLOT(start()) );
	connect( &m_dependenciesJob, SIGNAL(finished()), this, SIGNAL(finished()) );

	m_libraryPathsJob.start();
}


dpwalker/importsexportsjob.h
/***************************************************************************
 *   Copyright (C) 2007 by Bernhard Beschow                                *
 *   bbeschow (.at) cs.tu-berlin.de                                        *
 *   Copyright (C) 2020 by C6H6Vita                                        *
 *   c6h6 (.at) outlook.kr                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#ifndef IMPORTSEXPORTSJOB_H
#define IMPORTSEXPORTSJOB_H

#include <QMap>
#include <QProcess>
#include <QTextStream>

class QTreeWidget;
class QTreeWidgetItem;


class ImportsJob : public QObject
{
	Q_OBJECT
public:
	ImportsJob( const QString &file, QTreeWidget *pImports, const QMap<QString, QString> *pMap );

public slots:
	void start();

signals:
	void finished();

protected slots:
	void readLineStdout();

protected:
	QString m_file;
	QTreeWidget *m_pImports;
	const QMap<QString, QString> *m_pImportsMap;
	QProcess m_proc;
	QTextStream m_stream;
};

class ExportsJob : public QObject
{
	Q_OBJECT
public:
	ExportsJob( const QString &file, QTreeWidget *pExports, QMap<QString, QString> *pMap );

public slots:
	void start();

signals:
	void finished();

protected slots:
	void readLineStdout();

protected:
	QString m_file;
	QTreeWidget *m_pExports;
	QMap<QString, QString> *m_pExportsMap;
	QProcess m_proc;
	QTextStream m_stream;
};

/**
 * @author Bernhard Beschow <bbeschow (.at) cs.tu-berlin.de>
 */
class ImportsExportsJob : public QObject
{
	Q_OBJECT

public:
	ImportsExportsJob( QTreeWidgetItem *pItem, QTreeWidget *pImports, QTreeWidget *pExports );

signals:
	void finished();

protected:
	QTreeWidgetItem *m_pItem;
	QMap<QString, QString> m_map;
	ExportsJob m_exportsJob;
	ImportsJob m_importsJob;
};

#endif


dpwalker/importsexportsjob.cpp
/***************************************************************************
 *   Copyright (C) 2007 by Bernhard Beschow                                *
 *   bbeschow (.at) cs.tu-berlin.de                                        *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include "importsexportsjob.h"

#include <QRegExp>
#include <QTreeWidget>


QIcon iconFromType( const QString &type )
{
	if ( type == "F" )
	{
		return QIcon( ":/icons/function.png" );
	}

	if ( type == "O" )
	{
		return QIcon( ":/icons/object.png" );
	}

	return QIcon();
}


ExportsJob::ExportsJob( const QString &file, QTreeWidget *pExports, QMap<QString, QString> *pMap ) :
	m_file( file ),
	m_pExports( pExports ),
	m_pExportsMap( pMap ),
	m_proc(),
	m_stream( &m_proc )
{
	m_proc.setReadChannel( QProcess::StandardOutput );

	connect( &m_proc, SIGNAL(readyReadStandardOutput()), this, SLOT(readLineStdout()) );
	connect( &m_proc, SIGNAL(finished(int)), this, SLOT(readLineStdout()) );
}

void ExportsJob::start()
{
    const QStringList args = QStringList() << "-TC" << m_file;

	m_proc.start( "objdump", args );
}


void ExportsJob::readLineStdout()
{
	while ( m_proc.canReadLine() || !m_stream.atEnd() )
	{
		const QString line = m_stream.readLine();

		const QRegExp re( "^([0-9a-f]+)\\s(.)(.)(.)(.)(.)(.)(.)\\s(\\S+)\t([0-9a-f]+)\\s\\s?(\\s{11}|\\S+\\s*)\\s(\\S.*)$" );
		if ( re.indexIn( line ) >= 0 )
		{
            if ( re.cap( 9 ).compare( "*UND*" ) == 0 )
                continue;

			QTreeWidgetItem *pItem = new QTreeWidgetItem( m_pExports );

			pItem->setText( 0, re.cap( 12 ) );
			pItem->setText( 1, re.cap( 1 ) );
			pItem->setText( 2, re.cap( 2 ) );
			pItem->setText( 3, re.cap( 3 ) );
			pItem->setText( 4, re.cap( 4 ) );
			pItem->setText( 5, re.cap( 5 ) );
			pItem->setText( 6, re.cap( 6 ) );
			pItem->setText( 7, re.cap( 7 ) );
			pItem->setText( 8, re.cap( 9 ) );
			pItem->setText( 9, re.cap( 10 ) );
			pItem->setText( 10, re.cap( 11 ) );

			const QIcon icon = iconFromType( re.cap( 8 ) );
			pItem->setIcon( 0, icon );

			(*m_pExportsMap)[ re.cap( 12 ) ] = line;
		}
	}

	if ( m_proc.state() == QProcess::NotRunning )
	{
		emit finished();
	}
}


ImportsJob::ImportsJob( const QString &file, QTreeWidget *pImports, const QMap<QString, QString> *pMap ) :
	m_file( file ),
	m_pImports( pImports ),
	m_pImportsMap( pMap ),
	m_proc(),
	m_stream( &m_proc )
{
	m_proc.setReadChannel( QProcess::StandardOutput );

	connect( &m_proc, SIGNAL(readyReadStandardOutput()), this, SLOT(readLineStdout()) );
	connect( &m_proc, SIGNAL(finished(int)), this, SLOT(readLineStdout()) );
}


void ImportsJob::start()
{
	const QStringList args = QStringList() << "-TC" << m_file;

	m_proc.start( "objdump", args );
}


void ImportsJob::readLineStdout()
{
	while ( m_proc.canReadLine() || !m_stream.atEnd() )
	{
		const QString line = m_stream.readLine();

		const QRegExp re( "^[0-9a-f]+\\s......(.)\\s(\\S+)\t[0-9a-f]+\\s\\s?(\\s{11}|\\S+\\s*)\\s(\\S.*)$" );
		if ( re.indexIn( line ) >= 0 )
		{
			if ( re.cap( 2 ).compare( "*UND*" ) != 0 )
				continue;

			if ( m_pImportsMap->contains( re.cap( 4 ) ) )
			{
				QTreeWidgetItem *pItem = new QTreeWidgetItem( m_pImports );

				pItem->setText( 0, re.cap( 4 ) );

				const QIcon icon = iconFromType( re.cap( 1 ) );
				pItem->setIcon( 0, icon );
			}
		}
	}

	if ( m_proc.state() == QProcess::NotRunning )
	{
		emit finished();
	}
}

ImportsExportsJob::ImportsExportsJob( QTreeWidgetItem *pItem, QTreeWidget *pImports, QTreeWidget *pExports ) :
	m_pItem( pItem ),
	m_exportsJob( pItem->text( 1 ), pExports, &m_map ),
	m_importsJob( pItem->parent() ? pItem->parent()->text( 1 ) : "", pImports, &m_map )
{
	pImports->clear();
	pExports->clear();

	connect( &m_exportsJob, SIGNAL(finished()), &m_importsJob, SLOT(start()) );
	connect( &m_importsJob, SIGNAL(finished()), this, SIGNAL(finished()) );

	m_exportsJob.start();
}


dpwalker/dependsui.cpp
/***************************************************************************
 *   Copyright (C) 2007 by Bernhard Beschow                                *
 *   bbeschow (.at) cs.tu-berlin.de                                        *
 *   Copyright (C) 2020 by C6H6Vita                                        *
 *   c6h6 (.at) outlook.kr                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/




#include "dependsui.h"

#include "importsexportsjob.h"
#include "dependencyjob.h"

#include <QAction>
#include <QFileDialog>
#include <QHeaderView>
#include <QTreeWidgetItem>
#include <QMovie>
#include <QMessageBox>
#include <QIcon>

#include <unistd.h>
#include <iostream>
DependsUI::DependsUI( QWidget *parent, Qt::WindowFlags flags ) :
	QMainWindow( parent, flags ),
	m_ui()
{
	m_ui.setupUi( this );

    //Comment These lines enables QTreeWidget resizable
    /*m_ui.m_pListImports->header()->setSectionResizeMode( QHeaderView::ResizeToContents );
	m_ui.m_pListImports->setSortingEnabled( true );
    m_ui.m_pListExports->header()->setSectionResizeMode( QHeaderView::ResizeToContents );
	m_ui.m_pListExports->setSortingEnabled( true );
    m_ui.m_pTreeSharedObjects->header()->setSectionResizeMode( QHeaderView::ResizeToContents );*/

	connect( m_ui.m_pTreeSharedObjects, SIGNAL(itemExpanded(QTreeWidgetItem*)), this, SLOT(dependencyExpanded(QTreeWidgetItem*)) );
    //Comment this line and enables directly at *.ui file
    //connect( m_ui.m_pTreeSharedObjects, SIGNAL(itemClicked(QTreeWidgetItem*,int)), this, SLOT(dependencySelected(QTreeWidgetItem*)) );

    //set spinner setting
    gif=new QMovie(":/icons/spinner.gif");
    size_t lbSpinnerPxLen=m_ui.m_StatusbarWzt->height();
    QSize lbSpinnerPxSz(lbSpinnerPxLen,lbSpinnerPxLen);
    m_ui.m_lbSpinner->setMinimumSize(lbSpinnerPxSz);
    m_ui.m_lbSpinner->setMaximumSize(lbSpinnerPxSz);

    //put Statusbar Wizet to statusbar

    m_ui.m_BaseLayout->takeAt(1);
    m_ui.statusBar->addPermanentWidget(m_ui.m_StatusbarWzt,1);
    QFont fnt=m_ui.m_StatusbarWzt->font();
    fnt.setPointSize(fnt.pointSize()-2);
    m_ui.m_StatusbarWzt->setFont(fnt);

    setBusy(false);
}

DependsUI::~DependsUI()
{}

void DependsUI::fileExit()
{
	close();
}

void DependsUI::fileOpen()
{
	QFileDialog dlg( this );

	if ( ! m_dirPath.isEmpty() )
		dlg.setDirectory( QDir( m_dirPath ) );

	if ( dlg.exec() != QDialog::Accepted )
	{
		return;
	}

	m_dirPath = dlg.directory().path();

	const QStringList selectedFiles = dlg.selectedFiles();

	if ( selectedFiles.isEmpty() )
	{
		return;
	}

	openFile( selectedFiles.first() );
}

void DependsUI::openFile( const QString &file )
{
    setWindowTitle( tr( "DPWalker - [%1]" ).arg( QFileInfo( file ).fileName() ) );
	m_ui.m_pTreeSharedObjects->clear();
//	m_ui.m_pListSummarySharedObjects->clear();
	m_ui.m_pListImports->clear();
	m_ui.m_pListExports->clear();

	QTreeWidgetItem *pItem = new QTreeWidgetItem( m_ui.m_pTreeSharedObjects );
	pItem->setText( 0, QFileInfo( file ).fileName() );
	pItem->setText( 1, file );

	DependencyJobs *job = new DependencyJobs( pItem, this );
    connect(job,SIGNAL(finished()),job,SLOT(deleteLater()));
    connect(job,SIGNAL(finished()),this,SLOT(setRootItemExpanded()));
    connect(job,SIGNAL(finished()),this,SLOT(runCompleted()));
}

void DependsUI::dependencyExpanded( QTreeWidgetItem *pItem )
{
	for ( int i = 0; i < pItem->childCount(); ++i )
	{
		QTreeWidgetItem *childItem = pItem->child( i );
		DependencyJobs *job = new DependencyJobs( childItem, this );
        connect(job,SIGNAL(finished()),job,SLOT(deleteLater()));
        connect(job,SIGNAL(finished()),this,SLOT(runCompleted()));
	}
}

void DependsUI::dependencySelected( QTreeWidgetItem *pItem )
{
    setBusy(true);
	ImportsExportsJob *job = new ImportsExportsJob( pItem, m_ui.m_pListImports, m_ui.m_pListExports );
    connect(job,SIGNAL(finished()),job,SLOT(deleteLater()));
    connect(job,SIGNAL(finished()),this,SLOT(runCompleted()));
}

void DependsUI::runCompleted()
{
    setBusy(false);
}

void DependsUI::setRootItemExpanded()
{
    m_ui.m_pTreeSharedObjects->topLevelItem( 0 )->setExpanded( true );
}

void DependsUI::setBusy(bool isBusy)
{
    if(isBusy)
    {
        this->setCursor(Qt::WaitCursor);
        m_ui.m_lbSpinner->setMovie(gif);
        gif->start();
    }
    else
    {
        this->setCursor(Qt::ArrowCursor);
        gif->stop();
        m_ui.m_lbSpinner->clear();
    }
}
void DependsUI::aboutClicked()
{
    QIcon icon();
    QMessageBox::information(this,"About..","DPWalker v1.1.0");

}
